File: trunk/ClientServerInteraction.iOS/ClientServerInteraction.iOS/Serializer.m
//
//  Serializer.m
//  ClientServerInteraction.iOS
//
//  Created by Yuriy Pogrebnyak on 07.04.13.
//  Copyright (c) 2013 Alexander O. Taraymovich. All rights reserved.
//

#import "Serializer_FrameworkOnly.h"

@implementation Serializer

+(NSString*)prepareQueryString:(NSDictionary*)params {
    if (params == nil)
        return @"";
    NSString* result = @"";
    for (id key in params) {
        id value = [params objectForKey:key];
        result = [NSString stringWithFormat:[result isEqual: @""] ? @"%@%@=%@" : @"%@&%@=%@", result, key, value];
    }
    return result;
}

+(void)deserializeResponse:(NSDictionary*)json :(Class)responseClass :(void (^)(int code, id response, NSError* error, ServerResponseError* serverError))completionBlock {
    int responseCode = [[json objectForKey:@"responseCode"] intValue];
    if (responseCode == RESPONSE_ERROR) {
        completionBlock(SERVER_ERROR, nil, nil, [Serializer deserialize:[json objectForKey:@"error"] :[ServerResponseError class]]);
        return;
    }
    if (responseCode == OK) {
        completionBlock(OK, [Serializer deserialize:[json objectForKey:@"data"] :responseClass], nil, nil);
        return;
    }
    
}

+(id)deserialize:(id)obj :(Class)class {
    if (class == [User class])
        return [Serializer deserializeUser:obj];
    if (class == [NSNumber class])
        return [Serializer deserializeBool:obj];
    if (class == [AccessToken class])
        return [Serializer deserializeAccessToken:obj];
    if (class == [ServerResponseError class])
        return [Serializer deserializeServerError:obj];
    if (class == [Session class])
        return [Serializer deserializeSessionsArray:obj];
    if (class == [NSArray class])
        return obj;
    if (class == [NSDictionary class])
        return [Serializer deserializeTension:obj];
    return nil;
}

+(NSString*)serialize:(id)obj {
    if ([obj isKindOfClass:[User class]])
        return [Serializer serializeUser:obj];
    if ([obj isKindOfClass: [NSDictionary class]])
        return [Serializer serializeDictionary:obj];
    return nil;
}

+(NSString*)serializeUser:(User*) user {
    NSArray* objects = [NSArray arrayWithObjects:user.about, user.birthDate, user.description, user.department, user.diagnosis, user.email, user.firstName, user.lastName, user.height, user.password, user.sex, user.statusMessage, user.weight, nil];
    
    NSArray* keys = [NSArray arrayWithObjects:@"about", @"birthDate", @"description", @"department", @"diagnosis", @"email", @"firstName", @"lastName", @"height", @"password", @"sex", @"statusMessage", @"weight", nil];
    NSData* data = [NSJSONSerialization dataWithJSONObject:[NSDictionary dictionaryWithObjects:objects forKeys:keys] options:NSJSONWritingPrettyPrinted error:nil];
    NSString* jsonString = [[NSString alloc] initWithData:data encoding:NSUTF8StringEncoding];
    
    return jsonString;
}

+(NSString*)serializeDictionary:(NSDictionary*)dict {
    NSData* data = [NSJSONSerialization dataWithJSONObject:dict options:NSJSONWritingPrettyPrinted error:nil];
    NSString* jsonString = [[NSString alloc] initWithData:data encoding:NSUTF8StringEncoding];
    return jsonString;
}

+(NSDictionary*)deserializeTension:(NSArray*)array {
    NSMutableDictionary* tension = [[NSMutableDictionary alloc] init];
    for (id internalArray in array) {
        [tension setObject:internalArray[1] forKey:internalArray[0]];
    }
    return tension;
}

+(NSArray*)deserializeRates:(NSArray*)rates {
    return rates;
}

+(NSArray*)deserializeSessionsArray:(NSArray*) array {
    NSMutableArray* sessions = [[NSMutableArray alloc] init];
    for (id obj in array) {
        [sessions addObject:[Serializer deserializeSession:obj]];
    }
    return sessions;
}

+(Session*)deserializeSession:(NSDictionary*)dict
{
    Session* session = [Session alloc];
    session.sessionId = dict[@"id"];
    session.start = dict[@"start"];
    session.end = dict[@"end"];
    return session;
}

+(User*)deserializeUser:(NSDictionary*) dict {
    User* user = [User alloc];
    user.about = [dict objectForKey:@"about"];
    user.birthDate = [dict objectForKey:@"birthDate"];
    user.description = [dict objectForKey:@"description"];
    user.department = [dict objectForKey:@"department"];
    user.diagnosis = [dict objectForKey:@"diagnosis"];
    user.email = [dict objectForKey:@"email"];
    user.firstName = [dict objectForKey:@"firstName"];
    user.lastName = [dict objectForKey:@"lastName"];
    user.height = [dict objectForKey:@"height"];
    user.password = [dict objectForKey:@"password"];
    user.sex = [dict objectForKey:@"sex"];
    user.statusMessage = [dict objectForKey:@"statusMessage"];
    user.weight = [dict objectForKey:@"weight"];
    user.userId = [dict objectForKey:@"userId"];
    return user;
}

+(NSString*)serializeSession:(NSString*)email withPassword:(NSString*)password rates:(NSArray*) rates start:(long long) start create:(int)create {
    NSArray* objects = [NSArray arrayWithObjects:email, password, rates, [NSString stringWithFormat:@"%lld", start], [NSNumber numberWithInt:create], nil];
    
    NSArray* keys = [NSArray arrayWithObjects:@"email", @"password", @"rates", @"start", @"create", nil];
    NSData* data = [NSJSONSerialization dataWithJSONObject:[NSDictionary dictionaryWithObjects:objects forKeys:keys] options:NSJSONWritingPrettyPrinted error:nil];
    NSString* json = [[NSString alloc] initWithData:data encoding:NSUTF8StringEncoding];
    return json;
}

+(NSNumber*)deserializeBool:(NSString*)value {
    NSNumberFormatter * formatter = [[NSNumberFormatter alloc] init];
    [formatter setNumberStyle:NSNumberFormatterDecimalStyle];
    NSNumber * boolNumber = [formatter numberFromString:value];
    return boolNumber;
}

+(AccessToken*)deserializeAccessToken:(NSDictionary*)dict {
    AccessToken* accessToken = [AccessToken alloc];
    accessToken.token = [dict objectForKey:@"token"];
    accessToken.expiredDate = [dict objectForKey:@"expiredDate"];
    return accessToken;
}

+(ServerResponseError*)deserializeServerError:(NSDictionary*)dict {
    ServerResponseError* error = [ServerResponseError alloc];
    error.message = [dict objectForKey:@"message"];
    error.errorCode = [[dict objectForKey:@"code"] intValue];
    return error;
}

@end

File: ClientServerInteraction/src/com/omnihealth/client_server_interaction/Serializer.java
package com.omnihealth.client_server_interaction;

import android.util.Log;
import com.google.gson.Gson;
import org.json.JSONArray;
import org.json.JSONException;
import org.json.JSONObject;

import java.util.ArrayList;
import java.util.List;
import java.util.Map;

/**
 * Created with IntelliJ IDEA.
 * User: yuriy
 * Date: 30.03.13
 * Time: 23:47
 * To change this template use File | Settings | File Templates.
 */
public class Serializer {

    public static String makeQueryString(Map params) {
        String queryString = "";
        if (params != null) {
            for (Object key : params.keySet()) {
                queryString += String.format(queryString.equals("") ? "%s=%s" : "&%s=%s", key, params.get(key));
            }
        }
        return queryString;
    }

    private static List<Integer> deserializeRates(JSONArray data) throws JSONException {
        List<Integer> rates = new ArrayList<Integer>();
        for (int i = 0, length = data.length(); i < length; i++) {
            rates.add((Integer)data.get(i));
        }
        return rates;
    }

    private static List<double[]> deserializeTension(JSONArray data) throws JSONException {
        List<double[]> tension = new ArrayList<double[]>();
        for (int i = 0, length = data.length(); i < length; i++) {
            JSONArray item = (JSONArray)data.get(i);
            tension.add(new double[] {new Double(item.get(0).toString()), new Double(item.get(1).toString())});

        }
        return tension;
    }

    private static User deserializeUser(JSONObject json) {
        User user = new User();
        user.setAbout(json.optString("about"));
        user.setBirthDate(json.optString("birthDate"));
        user.setDepartment(json.optString("department"));
        user.setDescription(json.optString("description"));
        user.setDiagnosis(json.optString("diagnosis"));
        user.setEmail(json.optString("email"));
        user.setFirstName(json.optString("firstName"));
        user.setLastName(json.optString("lastName"));
        user.setHeight(json.optDouble("height"));
        user.setPassword(json.optString("password"));
        user.setSex(json.optInt("sex"));
        user.setStatusMessage(json.optString("statusMessage"));
        user.setWeight(json.optDouble("weight"));
        return user;
    }

    private static List<Session> deserializeSessionList(JSONArray data) throws JSONException {
        List<Session> sessions = new ArrayList<Session>();

        for (int i = 0, length = data.length(); i < length; i++) {
            sessions.add(deserializeSession((JSONObject)data.get(i)));
        }
        return sessions;
    }

    private static Session deserializeSession(JSONObject data) throws JSONException {
        Session session = new Session();
        session.setSessionId(data.getLong("id"));
        session.setStart(data.getLong("start"));
        session.setEnd(data.getLong("end"));
        return session;
    }

    private static AccessToken deserializeAccessToken(JSONObject json) throws JSONException {
        AccessToken at = new AccessToken();
        at.setToken(json.getString("token"));
        at.setExpiredDate(json.getLong("expiredDate"));
        return at;
    }

    public static String serialize(Object json) {
        Gson gson = new Gson();
        return gson.toJson(json);
    }

    public static Object deserialize(String json, int classId) throws ServerResponseError, JSONException {
        JSONObject jsonObject = new JSONObject(json);
        int responseCode = jsonObject.getInt("responseCode");
        if (responseCode == 1) {
            return deserializeData(jsonObject.get("data"), classId);
        } else {
            JSONObject error = new JSONObject(jsonObject.getString("error"));
            throw new ServerResponseError(error.getString("message"), error.getInt("code"));
        }
    }

    private static Object deserializeData(Object data, int classId) throws JSONException {
        switch (classId) {
            case ClassIds.INT_ID:
                return data;
            case ClassIds.USER_ID:
                return deserializeUser((JSONObject)data);
            case ClassIds.ACCESS_TOKEN_ID:
                return deserializeAccessToken((JSONObject)data);
            case ClassIds.LIST_OF_SESSONS_ID:
                return deserializeSessionList((JSONArray)data);
            case ClassIds.TENSION_ID:
                return deserializeTension((JSONArray)data);
            case ClassIds.RATES_ID:
                return deserializeRates((JSONArray)data);
        }
        return null;
    }





}
